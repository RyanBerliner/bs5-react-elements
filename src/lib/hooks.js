import {useEffect, useRef} from 'react';

/**
 * Hook to initialize a component, bind events, and expose bootstrap apis.
 *
 * @param {Object} Component Bootstrap component object (Modal, Tooltip, etc)
 * @param {Object} config (optional) Configuration object of the component
 * @param {Object} componentRef (optional) A ref to reference the component
 * @param {Object} domRef A ref pointing to the dom element of the component
 * @param {Map} events Mapping of bootstrap event names and function handlers
 */
export function useBootstrap(
    Component,
    config,
    componentRef,
    domRef,
    events,
) {
  const initialConfig = useRef(config);

  useEffect(() => {
    const component = Component.getInstance(domRef.current) ||
      new Component(domRef.current, initialConfig.current);
    if (componentRef) {
      componentRef.current = component;
    }

    return () => component.dispose();
  }, [Component, componentRef, domRef]);

  useEffect(() => {
    const el = domRef.current;
    events.forEach((value, key) => {
      el.addEventListener(key, value);
    });

    return () => {
      events.forEach((value, key) => {
        el.removeEventListener(key, value);
      });
    };
  }, [domRef, events]);
}
