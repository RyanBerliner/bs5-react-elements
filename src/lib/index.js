export {Alert} from './Alert.jsx';
export {Carousel} from './Carousel.jsx';
export {Dropdown} from './Dropdown.jsx';
export {Modal} from './Modal.jsx';
export {Offcanvas} from './Offcanvas.jsx';
export {Popover} from './Popover.jsx';
export {Tab} from './Tab.jsx';
export {Toast} from './Toast.jsx';
export {Tooltip} from './Tooltip.jsx';
