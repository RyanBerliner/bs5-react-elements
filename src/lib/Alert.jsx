import React, {useRef, useMemo, useCallback} from 'react';
import {unmountComponentAtNode} from 'react-dom';
import PropTypes from 'prop-types';
import {Alert} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.0/components/alerts/">Bootstrap alert component.</a>
 */
function AlertComponent({
  onClosed,
  onClose,
  container,
  component,
  children,
  ...props
}) {
  const componentElement = useRef();

  const _onClosed = useCallback((event) => {
    unmountComponentAtNode(container.current);

    if (onClosed) {
      onClosed(event);
    }
  }, [onClosed]);

  const events = useMemo(() => {
    return new Map([
      ['closed.bs.alert', _onClosed],
      ['close.bs.alert', onClose],
    ]);
  }, [_onClosed, onClose]);

  useBootstrap(Alert, undefined, component, componentElement, events);

  return (
    <div ref={componentElement} {...props}>
      {children}
    </div>
  );
}

AlertComponent.propTypes = {
  /**
   * Handler for the closed.bs.alert event
   */
  onClosed: PropTypes.func,
  /**
   * Handler for the close.bs.alert event
   */
  onClose: PropTypes.func,
  /**
   * React ref refering to the container the element is wrapped in
   */
  container: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]).isRequired(),
  /**
   * React ref that will be assigned the component instance
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Alert contents
   */
  children: PropTypes.node,
};

AlertComponent.displayName = 'Alert';

export {AlertComponent as Alert};
