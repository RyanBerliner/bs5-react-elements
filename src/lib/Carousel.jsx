import React, {useRef, useMemo} from 'react';
import PropTypes from 'prop-types';
import {Carousel} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.2/components/carousel/">Bootstrap carousel component.</a>
 */
function CarouselComponent({
  onSlid,
  onSlide,
  component,
  config,
  children,
  ...props
}) {
  const componentElement = useRef();

  const events = useMemo(() => {
    return new Map([
      ['slid.bs.carousel', onSlid],
      ['slide.bs.carousel', onSlide],
    ]);
  }, [onSlid, onSlide]);

  useBootstrap(Carousel, config, component, componentElement, events);

  return (
    <div ref={componentElement} {...props}>
      {children}
    </div>
  );
}

CarouselComponent.propTypes = {
  /**
   * Handler for the slid.bs.carousel event
   * @since 0.2.0
   */
  onSlid: PropTypes.func,
  /**
   * Handler for the slide.bs.carousel event
   * @since 0.2.0
   */
  onSlide: PropTypes.func,
  /**
   * React ref that will be assigned the component instance
   * @since 0.2.0
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Component configuration
   * @since 0.2.0
   */
  config: PropTypes.object,
  /**
   * Dropdown contents
   * @since 0.2.0
   */
  children: PropTypes.node,
};

CarouselComponent.displayName = 'Carousel';

export {CarouselComponent as Carousel};
