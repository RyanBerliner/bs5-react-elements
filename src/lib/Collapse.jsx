import React, {useRef, useMemo} from 'react';
import PropTypes from 'prop-types';
import {Collapse} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.0/components/collapse/">Bootstrap collapse component.</a>
 */
function CollapseComponent({
  onShown,
  onShow,
  onHidden,
  onHide,
  component,
  config,
  children,
  ...props
}) {
  const componentElement = useRef();

  const events = useMemo(() => {
    return new Map([
      ['shown.bs.collapse', onShown],
      ['show.bs.collapse', onShow],
      ['hidden.bs.collapse', onHidden],
      ['hide.bs.collapse', onHide],
    ]);
  }, [onShown, onShow, onHidden, onHide]);

  useBootstrap(Collapse, config, component, componentElement, events);

  return (
    <div ref={componentElement} {...props}>
      {children}
    </div>
  );
}

CollapseComponent.propTypes = {
  /**
   * Handler for the shown.bs.collapse event
   */
  onShown: PropTypes.func,
  /**
   * Handler for the show.bs.collapse event
   */
  onShow: PropTypes.func,
  /**
   * Handler for the hidden.bs.collapse event
   */
  onHidden: PropTypes.func,
  /**
   * Handler for the hide.bs.collapse event
   */
  onHide: PropTypes.func,
  /**
   * React ref that will be assigned the component instance
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Component configuration
   */
  config: PropTypes.object,
  /**
   * Collapse contents
   */
  children: PropTypes.node,
};

CollapseComponent.displayName = 'Collapse';

export {CollapseComponent as Collapse};
