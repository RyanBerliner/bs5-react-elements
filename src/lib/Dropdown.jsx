import React, {useRef, useMemo} from 'react';
import PropTypes from 'prop-types';
import {Dropdown} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.3/components/dropdowns/">Bootstrap dropdown component.</a>
 */
function DropdownComponent({
  onShown,
  onShow,
  onHidden,
  onHide,
  component,
  config,
  children,
  as: ElementType,
  ...props
}) {
  const componentElement = useRef();

  const events = useMemo(() => {
    return new Map([
      ['shown.bs.dropdown', onShown],
      ['show.bs.dropdown', onShow],
      ['hidden.bs.dropdown', onHidden],
      ['hide.bs.dropdown', onHide],
    ]);
  }, [onShown, onShow, onHidden, onHide]);

  useBootstrap(Dropdown, config, component, componentElement, events);

  return (
    <ElementType ref={componentElement} {...props}>
      {children}
    </ElementType>
  );
}

DropdownComponent.propTypes = {
  /**
   * Handler for the shown.bs.dropdown event
   * @since 0.2.0
   */
  onShown: PropTypes.func,
  /**
   * Handler for the show.bs.dropdown event
   * @since 0.2.0
   */
  onShow: PropTypes.func,
  /**
   * Handler for the hidden.bs.dropdown event
   * @since 0.2.0
   */
  onHidden: PropTypes.func,
  /**
   * Handler for the hide.bs.dropdown event
   * @since 0.2.0
   */
  onHide: PropTypes.func,
  /**
   * React ref that will be assigned the component instance
   * @since 0.2.0
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Component configuration
   * @since 0.2.0
   */
  config: PropTypes.object,
  /**
   * Dropdown contents
   * @since 0.2.0
   */
  children: PropTypes.node,
  /**
   * The element type of the dropdown
   * @since 0.2.0
   */
  as: PropTypes.elementType,
};

DropdownComponent.defaultProps = {
  as: 'button',
};

DropdownComponent.displayName = 'Dropdown';

export {DropdownComponent as Dropdown};
