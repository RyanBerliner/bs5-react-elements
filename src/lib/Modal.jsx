import React, {useRef, useMemo} from 'react';
import PropTypes from 'prop-types';
import {Modal} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.3/components/modal/">Bootstrap modal component.</a>
 */
function ModalComponent({
  onShown,
  onShow,
  onHidden,
  onHide,
  onHidePrevented,
  component,
  config,
  children,
  ...props
}) {
  const componentElement = useRef();

  const events = useMemo(() => {
    return new Map([
      ['shown.bs.modal', onShown],
      ['show.bs.modal', onShow],
      ['hidden.bs.modal', onHidden],
      ['hide.bs.modal', onHide],
      ['hidePrevented.bs.modal', onHidePrevented],
    ]);
  }, [onShown, onShow, onHidden, onHide, onHidePrevented]);

  useBootstrap(Modal, config, component, componentElement, events);

  return (
    <div ref={componentElement} {...props}>
      {children}
    </div>
  );
}

ModalComponent.propTypes = {
  /**
   * Handler for the shown.bs.modal event
   * @since 0.1.0
   */
  onShown: PropTypes.func,
  /**
   * Handler for the show.bs.modal event
   * @since 0.1.0
   */
  onShow: PropTypes.func,
  /**
   * Handler for the hidden.bs.modal event
   * @since 0.1.0
   */
  onHidden: PropTypes.func,
  /**
   * Handler for the hide.bs.modal event
   * @since 0.1.0
   */
  onHide: PropTypes.func,
  /**
   * Handler for the hidePrevented.bs.modal event
   * @since 0.1.0
   */
  onHidePrevented: PropTypes.func,
  /**
   * React ref that will be assigned the component instance
   * @since 0.1.0
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Component configuration
   * @since 0.1.0
   */
  config: PropTypes.object,
  /**
   * Modal contents
   * @since 0.1.0
   */
  children: PropTypes.node,
};

ModalComponent.displayName = 'Modal';

export {ModalComponent as Modal};
