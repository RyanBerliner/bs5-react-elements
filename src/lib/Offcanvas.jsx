import React, {useRef, useMemo} from 'react';
import PropTypes from 'prop-types';
import {Offcanvas} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.0/components/offcanvas/">Bootstrap offcanvas component.</a>
 */
function OffcanvasComponent({
  onShown,
  onShow,
  onHidden,
  onHide,
  component,
  config,
  children,
  ...props
}) {
  const componentElement = useRef();

  const events = useMemo(() => {
    return new Map([
      ['shown.bs.offcanvas', onShown],
      ['show.bs.offcanvas', onShow],
      ['hidden.bs.offcanvas', onHidden],
      ['hide.bs.offcanvas', onHide],
    ]);
  }, [onShown, onShow, onHidden, onHide]);

  useBootstrap(Offcanvas, config, component, componentElement, events);

  return (
    <div ref={componentElement} {...props}>
      {children}
    </div>
  );
}

OffcanvasComponent.propTypes = {
  /**
   * Handler for the shown.bs.offcanvas event
   */
  onShown: PropTypes.func,
  /**
   * Handler for the show.bs.offcanvas event
   */
  onShow: PropTypes.func,
  /**
   * Handler for the hidden.bs.offcanvas event
   */
  onHidden: PropTypes.func,
  /**
   * Handler for the hide.bs.offcanvas event
   */
  onHide: PropTypes.func,
  /**
   * React ref that will be assigned the component instance
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Component configuration
   */
  config: PropTypes.object,
  /**
   * Offcanvas contents
   */
  children: PropTypes.node,
};

OffcanvasComponent.displayName = 'Offcanvas';

export {OffcanvasComponent as Offcanvas};
