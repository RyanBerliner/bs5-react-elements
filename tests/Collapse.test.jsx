import React, {useEffect, useRef} from 'react';
import {Collapse} from '../src/lib/Collapse';
import {render, screen, fireEvent} from '@testing-library/react';

/**
 * A boilerplate collapse, collapsed on load, with a button to open it.
 */
function TestCollapse(props) {
  return <>
    <button
      data-bs-toggle="collapse"
      data-bs-target="#test-collapse"
      aria-expanded="false"
      aria-controls="test-collapse"
    >
      Toggle
    </button>
    <Collapse id="test-collapse" className="collapse" {...props}>
      Lorem ipsum dolor.
    </Collapse>
  </>;
}

/**
 * Toggles the test collapse.
 */
function toggle() {
  fireEvent.click(screen.getByText('Toggle'));
}

// Timeouts are used to prevent disposing of bootstrap components that are in
// the middle of a transition. This is a bootstrap bug. TODO: Submit an issue.

describe('events bound to props', () => {
  test('show.bs.collapse event bound to onShow prop', (done) => {
    /**
     * show.bs.collapse event handler.
     *
     * @param {Object} event
     */
    function onShow(event) {
      expect(event).toBeTruthy();
      setTimeout(() => done(), 500);
    }

    render(<TestCollapse data-bs-toggle="false" onShow={onShow} />);
    toggle();
  });

  test('shown.bs.collapse event bound to onShown prop', (done) => {
    /**
     * shown.bs.collapse event handler.
     *
     * @param {Object} event
     */
    function onShown(event) {
      expect(event).toBeTruthy();
      setTimeout(() => done(), 500);
    }

    render(<TestCollapse data-bs-toggle="false" onShown={onShown} />);
    toggle();
  });

  test('hide.bs.collapse event bound to onHide prop', (done) => {
    /**
     * hide.bs.collapse event handler.
     *
     * @param {Object} event
     */
    function onHide(event) {
      expect(event).toBeTruthy();
      setTimeout(() => done(), 500);
    }

    render(
        <TestCollapse
          data-bs-toggle="false"
          onShown={toggle}
          onHide={onHide}
        />,
    );
    toggle();
  });

  test('hidden.bs.collapse event bound to onHidden prop', (done) => {
    /**
     * hidden.bs.collapse event handler.
     *
     * @param {Object} event
     */
    function onHidden(event) {
      expect(event).toBeTruthy();
      setTimeout(() => done(), 500);
    }

    render(
        <TestCollapse
          data-bs-toggle="false"
          onShown={toggle}
          onHidden={onHidden}
        />,
    );
    toggle();
  });
});

describe('bootstrap apis', () => {
  test('component object exposed', (done) => {
    /**
     * A collapse that will call the bootstrap show() method immediately on
     * mount.
     */
    function TestComponent(props) {
      const component = useRef();
      useEffect(() => component.current.show());
      return <TestCollapse
        data-bs-toggle="false"
        component={component}
        {...props}
      />;
    }

    /**
     * show.bs.collapse event handler.
     */
    function onShow() {
      setTimeout(() => done(), 500);
    }

    render(<TestComponent onShow={onShow} />);
  });

  test('config prop passed to component', () => {
    const {container} = render(<TestCollapse config={{toggle: false}} />);
    expect(container.querySelector('.collapsing')).toBeNull();
  });
});

describe('markup', () => {
  test('minimum markup renders as expected', () => {
    const {container} = render(<Collapse />);
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('places children in correct element', () => {
    const {container} = render(<Collapse>child</Collapse>);
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('support additional attributes, place them correctly', () => {
    const {container} = render(<Collapse data-lorem="ipsum" />);
    expect(container.innerHTML).toMatchSnapshot();
  });
});
