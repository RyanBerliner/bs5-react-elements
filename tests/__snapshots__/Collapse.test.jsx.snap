// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markup minimum markup renders as expected 1`] = `"<div class="collapsing" style="height: 0px;"></div>"`;

exports[`markup places children in correct element 1`] = `"<div class="collapsing" style="height: 0px;">child</div>"`;

exports[`markup support additional attributes, place them correctly 1`] = `"<div data-lorem="ipsum" class="collapsing" style="height: 0px;"></div>"`;
